File: ios/cross_bluetooth_api/Sources/cross_bluetooth_api/Commands/GetCharacteristicCommand.swift
import Foundation
import Flutter
import CoreBluetooth

public class GetCharacteristicCommand: BaseCommand, DeviceManagerDelegate {
    private let manager: BluetoothManager
    private var deviceManager: DeviceManager?
    private var continuation: CheckedContinuation<Void, Never>?
    private var targetDeviceId: String?
    private var targetServiceUUID: CBUUID?
    private var targetCharacteristicUUID: CBUUID?

    init(
        manager: BluetoothManager,
        arguments: [String: AnyObject]?,
        pendingResult: @escaping FlutterResult
    ) {
        self.manager = manager
        super.init(id: .getCharacteristic, arguments: arguments, pendingResult: pendingResult)
    }

    deinit {
        if let deviceManager {
            deviceManager.removeDelegate(self)
        }
    }

    override func execute() async {
        return await withCheckedContinuation { continuation in
            self.continuation = continuation

            guard
                let deviceId = arguments["deviceId"] as? String,
                let serviceUUIDString = arguments["serviceUUID"] as? String,
                let characteristicUUIDString = arguments["characteristic"] as? String,
                let deviceManager = manager.deviceManager(for: deviceId)
            else {
                pendingResult(FlutterError.networkError())
                continuation.resume()
                return
            }

            self.deviceManager = deviceManager
            deviceManager.addDelegate(self)

            let serviceUUID = CBUUID(string: serviceUUIDString)
            guard let service = deviceManager.getService(serviceUUID) else {
                pendingResult(FlutterError.notFoundError())
                continuation.resume()
                return
            }

            targetDeviceId = deviceId
            targetServiceUUID = serviceUUID
            targetCharacteristicUUID = CBUUID(string: characteristicUUIDString)
            deviceManager.discoverCharacteristics([targetCharacteristicUUID!], for: service)
        }
    }

    public func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        defer {
            continuation?.resume()
        }

        if error != nil {
            pendingResult(FlutterError.notFoundError())
            return
        }

        guard
            peripheral.identifier.uuidString == targetDeviceId,
            service.uuid == targetServiceUUID,
            let characteristicUUID = targetCharacteristicUUID,
            let characteristic = service.characteristics?.first(where: { $0.uuid == characteristicUUID })
        else {
            pendingResult(FlutterError.notFoundError())
            return
        }

        pendingResult(characteristic.toMap())
    }
}

File: android/src/main/kotlin/io/github/kokemus/cross_bluetooth_api/commands/GetCharacteristicCommand.kt
package io.github.kokemus.cross_bluetooth_api.commands

import io.github.kokemus.cross_bluetooth_api.extensions.notFoundError
import io.github.kokemus.cross_bluetooth_api.extensions.toMap
import io.github.kokemus.cross_bluetooth_api.services.BluetoothManager
import io.flutter.plugin.common.MethodChannel

class GetCharacteristicCommand(
    private val manager: BluetoothManager,
    arguments: Map<String, Any>,
    pendingResult: MethodChannel.Result
) : BaseCommand(CommandId.GET_CHARACTERISTIC, arguments, pendingResult) {
    override suspend fun execute() {
        val deviceId = arguments["deviceId"] as? String
        val serviceUUID = arguments["serviceUUID"] as? String
        val characteristicUUID = arguments["characteristic"] as? String
        if (deviceId == null || serviceUUID == null || characteristicUUID == null) {
            pendingResult.notFoundError()
            return
        }

        val characteristic = manager.deviceManager(deviceId)?.getCharacteristic(serviceUUID, characteristicUUID)
        if (characteristic != null) {
            pendingResult.success(characteristic.toMap())
        } else {
            pendingResult.notFoundError()
        }
    }
}
